File: src/types/equipment/item/Armor.ts
import * as DB from "tsondb/schema/dsl"
import { ArmorIdentifier, ArmorTypeIdentifier } from "../../_Identifier.js"
import { NestedTranslationMap } from "../../Locale.js"
import { Errata } from "../../source/_Erratum.js"
import { src } from "../../source/_PublicationRef.js"
import { ComplexComplexity, Cost, RestrictedTo, Weight } from "./_Item.js"

export const Armor = DB.Entity(import.meta.url, {
  name: "Armor",
  namePlural: "Armor",
  type: () =>
    DB.Object({
      cost: DB.Required({
        comment: "The cost in silverthalers.",
        type: DB.IncludeIdentifier(Cost),
      }),
      weight: DB.Required({
        comment: "The weight in kg.",
        type: DB.IncludeIdentifier(Weight),
      }),
      complexity: DB.Optional({
        comment: "The complexity of crafting the item.",
        type: DB.IncludeIdentifier(ArmorComplexity),
      }),
      protection: DB.Required({
        comment: "The PRO value.",
        type: DB.IncludeIdentifier(Protection),
      }),
      encumbrance: DB.Required({
        comment: "The ENC value.",
        type: DB.IncludeIdentifier(Encumbrance),
      }),
      has_additional_penalties: DB.Required({
        comment: "Does the armor have additional penalties (MOV -1, INI -1)?",
        type: DB.IncludeIdentifier(HasAdditionalPenalties),
      }),
      armor_type: DB.Required({
        comment: "The armor type..",
        type: ArmorTypeIdentifier(),
      }),
      hit_zone: DB.Optional({
        comment: "Specify if armor is only available for a specific hit zone.",
        type: DB.IncludeIdentifier(HitZone),
      }),
      restrictedTo: DB.Optional({
        comment:
          "Define if during character creation this weapon can only be bought by a specific subset of characters.",
        type: DB.IncludeIdentifier(RestrictedTo),
      }),
      src,
      translations: NestedTranslationMap(
        DB.Required,
        "Armor",
        DB.Object({
          name: DB.Required({
            comment: "The item’s name.",
            type: DB.String({ minLength: 1 }),
          }),
          secondary_name: DB.Optional({
            comment: "An auxiliary name or label of the item, if available.",
            type: DB.String({ minLength: 1 }),
          }),
          note: DB.Optional({
            comment: "Note text.",
            type: DB.String({ minLength: 1, isMarkdown: true }),
          }),
          rules: DB.Optional({
            comment: "Special rules text.",
            type: DB.String({ minLength: 1, isMarkdown: true }),
          }),
          advantage: DB.Optional({
            comment: "The armor advantage text.",
            type: DB.String({ minLength: 1, isMarkdown: true }),
          }),
          disadvantage: DB.Optional({
            comment: "The armor disadvantage text.",
            type: DB.String({ minLength: 1, isMarkdown: true }),
          }),
          errata: DB.Optional({
            type: DB.IncludeIdentifier(Errata),
          }),
        }),
      ),
    }),
  instanceDisplayName: {},
  uniqueConstraints: [
    {
      entityMapKeyPath: "translations",
      keyPathInEntityMap: "name",
    },
  ],
})

export const SecondaryArmor = DB.TypeAlias(import.meta.url, {
  name: "SecondaryArmor",
  type: () =>
    DB.Object({
      protection: DB.Required({
        comment: "The PRO value.",
        type: DB.IncludeIdentifier(Protection),
      }),
      encumbrance: DB.Required({
        comment: "The ENC value.",
        type: DB.IncludeIdentifier(Encumbrance),
      }),
      has_additional_penalties: DB.Required({
        comment: "Does the armor have additional penalties (MOV -1, INI -1)?",
        type: DB.IncludeIdentifier(HasAdditionalPenalties),
      }),
      armor_type: DB.Required({
        comment: "The armor type..",
        type: ArmorTypeIdentifier(),
      }),
      hit_zone: DB.Optional({
        comment: "Specify if armor is only available for a specific hit zone.",
        type: DB.IncludeIdentifier(HitZone),
      }),
      restrictedTo: DB.Optional({
        comment:
          "Define if during character creation this armor can only be bought by a specific subset of characters.",
        type: DB.IncludeIdentifier(RestrictedTo),
      }),
      translations: NestedTranslationMap(
        DB.Optional,
        "SecondaryArmor",
        DB.Object(
          {
            advantage: DB.Optional({
              comment: "The armor advantage text.",
              type: DB.String({ minLength: 1, isMarkdown: true }),
            }),
            disadvantage: DB.Optional({
              comment: "The armor disadvantage text.",
              type: DB.String({ minLength: 1, isMarkdown: true }),
            }),
          },
          { minProperties: 1 },
        ),
      ),
    }),
})

const ArmorComplexity = DB.Enum(import.meta.url, {
  name: "ArmorComplexity",
  comment: "The complexity of crafting the armor.",
  values: () => ({
    Primitive: DB.EnumCase({ type: null }),
    Simple: DB.EnumCase({ type: null }),
    Complex: DB.EnumCase({ type: DB.IncludeIdentifier(ComplexComplexity) }),
    Various: DB.EnumCase({ type: null }),
  }),
})

export const Protection = DB.TypeAlias(import.meta.url, {
  name: "Protection",
  comment: "The PRO value.",
  type: () => DB.Integer({ minimum: 0 }),
})

export const Encumbrance = DB.TypeAlias(import.meta.url, {
  name: "Encumbrance",
  comment: "The ENC value.",
  type: () => DB.Integer({ minimum: 0 }),
})

export const HasAdditionalPenalties = DB.TypeAlias(import.meta.url, {
  name: "HasAdditionalPenalties",
  comment: "Does the armor have additional penalties (MOV -1, INI -1)?",
  type: () => DB.Boolean(),
})

const HitZone = DB.Enum(import.meta.url, {
  name: "HitZone",
  comment: "Specify if armor is only available for a specific hit zone.",
  values: () => ({
    Head: DB.EnumCase({ type: DB.IncludeIdentifier(HeadHitZone) }),
    Torso: DB.EnumCase({ type: null }),
    Arms: DB.EnumCase({ type: null }),
    Legs: DB.EnumCase({ type: null }),
  }),
})

const HeadHitZone = DB.TypeAlias(import.meta.url, {
  name: "HeadHitZone",
  type: () =>
    DB.Object({
      combination_possibilities: DB.Optional({
        comment:
          "In some cases, multiple armors for the same hit zone can be combined. They're listed at the item that can be combined with others.",
        type: DB.IncludeIdentifier(HeadHitZoneCombinationPossibilities),
      }),
    }),
})

const HeadHitZoneCombinationPossibilities = DB.TypeAlias(import.meta.url, {
  name: "HeadHitZoneCombinationPossibilities",
  type: () =>
    DB.Object({
      armors: DB.Required({
        comment: "A list of armors that can be combined with this armor.",
        type: DB.Array(ArmorIdentifier(), { minItems: 1, uniqueItems: true }),
      }),
      protection: DB.Optional({
        comment:
          "The PRO value that is added to the PRO value of the other armor instead of adding the normale PRO value.",
        type: DB.Integer({ minimum: 0 }),
      }),
    }),
})

File: Sources/OptolithDatabaseSchema/Types/Equipment/Items/Armor.swift
import FileDB

@Model
public struct Armor {
    /// The cost in silverthalers.
    let cost: Cost

    /// The weight in kg.
    let weight: Weight

    /// The complexity of crafting the item.
    let complexity: Complexity

    /// The PRO value.
    let protection: Protection

    /// The ENC value.
    let encumbrance: Encumbrance

    /// Does the armor have additional penalties (MOV -1, INI -1)?
    let has_additional_penalties: HasAdditionalPenalties

    /// The armor type..
    @Relationship(ArmorType.self)
    let armor_type: ArmorType.ID

    /// Specify if armor is only available for a specific hit zone.
    let hit_zone: HitZone?

    /// The publications where you can find the entry.
    @MinItems(1)
    let src: [PublicationRef]

    /// All translations for the entry, identified by IETF language tag (BCP47).
    @Relationship(Locale.self)
    let translations: [String: Translation]

    @Embedded
    struct Translation {  // ArmorTranslation
        /// The item’s name.
        @MinLength(1)
        let name: String

        /// An auxiliary name or label of the item, if available.
        @MinLength(1)
        let secondary_name: String?

        /// Note text.
        @MinLength(1)
        @Markdown
        let note: String?

        /// Special rules text.
        @MinLength(1)
        @Markdown
        let rules: String?

        /// The armor advantage text.
        @MinLength(1)
        @Markdown
        let advantage: String?

        /// The armor disadvantage text.
        @MinLength(1)
        @Markdown
        let disadvantage: String?

        /// A list of errata for the entry in the specific language.
        @MinItems(1)
        let errata: [Erratum]?
    }

    public static let namePlural = "Armor"
}

@Embedded
public struct SecondaryArmor {
    /// The PRO value.
    let protection: Protection

    /// The ENC value.
    let encumbrance: Encumbrance

    /// Does the armor have additional penalties (MOV -1, INI -1)?
    let has_additional_penalties: HasAdditionalPenalties

    /// The armor type..
    @Relationship(ArmorType.self)
    let armor_type: ArmorType.ID

    /// Specify if armor is only available for a specific hit zone.
    let hit_zone: HitZone?

    /// All translations for the entry, identified by IETF language tag (BCP47).
    @Relationship(Locale.self)
    let translations: [String: Translation]?

    @Embedded
    @MinProperties(1)
    struct Translation {  // SecondaryArmorTranslation
        /// The armor advantage text.
        @MinLength(1)
        @Markdown
        let advantage: String?

        /// The armor disadvantage text.
        @MinLength(1)
        @Markdown
        let disadvantage: String?
    }
}

/// The PRO value.
@TypeAlias
public struct Protection {
    @Minimum(0)
    let type: Int
}

/// The ENC value.
@TypeAlias
public struct Encumbrance {
    @Minimum(0)
    let type: Int
}

/// Does the armor have additional penalties (MOV -1, INI -1)?
@TypeAlias
public struct HasAdditionalPenalties {
    let type: Bool
}

/// Specify if armor is only available for a specific hit zone.
@ModelEnum
public enum HitZone {
    case head(HeadHitZone)
    case torso
    case arms
    case legs
}

@Embedded
public struct HeadHitZone {
    /// In some cases, multiple armors for the same hit zone can be combined. They're listed at the item that can be combined with others.
    let combination_possibilities: HeadHitZoneCombinationPossibilities?
}

@Embedded
public struct HeadHitZoneCombinationPossibilities {
    /// A list of armors that can be combined with this armor.
    @MinItems(1)
    @UniqueItems
    @Relationship(Armor.self)
    let armors: [Armor.ID]

    /// The PRO value that is added to the PRO value of the other armor instead of adding the normale PRO value.
    @Minimum(0)
    let protection: Int?
}
